lib/multi.c

```diff
@@ -1049,46 +1049,69 @@ static int protocol_getsock(struct Curl_easy *data,
   return Curl_conn_get_select_socks(data, FIRSTSOCKET, socks);
 }
 
-/* returns bitmapped flags for this handle and its sockets. The 'socks[]'
-   array contains MAX_SOCKSPEREASYHANDLE entries. */
-static int multi_getsock(struct Curl_easy *data,
-                         curl_socket_t *socks)
+/* Initializes `poll_set` with the current socket poll actions needed
+ * for transfer `data`. */
+static void multi_getsock(struct Curl_easy *data,
+                          struct easy_poll_set *poll_set)
 {
   struct connectdata *conn = data->conn;
+  int actions;
   /* The no connection case can happen when this is called from
      curl_multi_remove_handle() => singlesocket() => multi_getsock().
   */
+  poll_set->num = 0;
   if(!conn)
-    return 0;
+    return;
 
   switch(data->mstate) {
   default:
-    return 0;
+    return;
 
   case MSTATE_RESOLVING:
-    return Curl_resolv_getsock(data, socks);
+    actions = Curl_resolv_getsock(data, poll_set->sockets);
+    break;
 
   case MSTATE_PROTOCONNECTING:
   case MSTATE_PROTOCONNECT:
-    return protocol_getsock(data, conn, socks);
+    actions = protocol_getsock(data, conn, poll_set->sockets);
+    break;
 
   case MSTATE_DO:
   case MSTATE_DOING:
-    return doing_getsock(data, conn, socks);
+    actions = doing_getsock(data, conn, poll_set->sockets);
+    break;
 
   case MSTATE_TUNNELING:
   case MSTATE_CONNECTING:
-    return Curl_conn_get_select_socks(data, FIRSTSOCKET, socks);
+    actions = Curl_conn_get_select_socks(data, FIRSTSOCKET, poll_set->sockets);
+    break;
 
   case MSTATE_DOING_MORE:
-    return domore_getsock(data, conn, socks);
+    actions = domore_getsock(data, conn, poll_set->sockets);
+    break;
 
   case MSTATE_DID: /* since is set after DO is completed, we switch to
                         waiting for the same as the PERFORMING state */
   case MSTATE_PERFORMING:
-    return Curl_single_getsock(data, conn, socks);
+    actions = Curl_single_getsock(data, conn, poll_set->sockets);
+    break;
   }
 
+  if(actions) {
+    int i;
+    for(i = 0; i < MAX_SOCKSPEREASYHANDLE; ++i) {
+      if(!(actions & GETSOCK_MASK_RW(i)) ||
+         !VALID_SOCK((poll_set->sockets[i]))) {
```

Comment:
Yes, the changed `reset` will take care of that. But the primary fault lies in a `getsock()` handler implementation that returns bits set for an invalid socket.

Since this was never ASSERTed before, just ignored, the changed code followed that.